File: android/app/src/main/java/com/jaewonbaek/wgfilesender/data/NetInfo.kt
package com.jaewonbaek.wgfilesender.data

import java.net.Inet4Address
import java.net.NetworkInterface

object NetInfo {
    /** IPv4 addresses on tunnel interfaces (tun*) — best-effort for showing this device's
     *  WireGuard address so it can be read off when pairing from the other side. */
    fun tunnelIPv4Addresses(): List<String> = runCatching {
        NetworkInterface.getNetworkInterfaces().asSequence()
            .filter { it.isUp && !it.isLoopback && it.name.startsWith("tun") }
            .flatMap { it.inetAddresses.asSequence() }
            .filterIsInstance<Inet4Address>()
            .mapNotNull { it.hostAddress }
            .toList()
    }.getOrDefault(emptyList())
}

File: macos/Sources/WGFileSender/Core/NetInfo.swift
import Foundation

enum NetInfo {
    /// IPv4 addresses on tunnel interfaces (utun*) — best-effort for showing this device's
    /// WireGuard address so the user can read it off when pairing from the other side.
    static func tunnelIPv4Addresses() -> [String] {
        var out: [String] = []
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let first = ifaddr else { return out }
        defer { freeifaddrs(ifaddr) }
        for ptr in sequence(first: first, next: { $0.pointee.ifa_next }) {
            let flags = Int32(ptr.pointee.ifa_flags)
            guard flags & IFF_UP != 0, flags & IFF_LOOPBACK == 0 else { continue }
            guard let sa = ptr.pointee.ifa_addr, sa.pointee.sa_family == UInt8(AF_INET) else { continue }
            let name = String(cString: ptr.pointee.ifa_name)
            guard name.hasPrefix("utun") else { continue }
            var host = [CChar](repeating: 0, count: Int(NI_MAXHOST))
            if getnameinfo(sa, socklen_t(sa.pointee.sa_len), &host, socklen_t(host.count),
                           nil, 0, NI_NUMERICHOST) == 0 {
                let addr = String(cString: host)
                if !addr.isEmpty { out.append(addr) }
            }
        }
        return out
    }
}
